### ISSUE TITLE
TypeScript: Select plugin issue

### ISSUE THREAD

**Issue by nWidart** [rematch/rematch#524]
Hello,

In a typescript project, I'm trying to use the select plugin, with the re-select api.

In the Typescript example, there's this: https://github.com/rematch/rematch/blob/master/examples/ts/count/src/models/sharks.ts#L20
```js
	selectors: slice => ({
		total() {
			return slice
		},
	}),
```

However, if I use the same structure I'm getting the following error:

```
error TS2717: Subsequent property declarations must have the same type.  Property 'selectors' must be of type '{ [key: string]: (state: SS, ...args: any[]) => any; }', but here has type 'ModelSelectorsConfig<S>'.
```

It looks like the module overwrite, https://github.com/rematch/rematch/blob/master/plugins/select/src/typings.d.ts#L68, wasn't taken into account?

**This happens when requiring the rematch/select plugin, even without any personal code.**

***

Also, when trying to use my selector, I'm getting this:

Selector (not using re-select):

```js
  selectors: {
    allEnabledEngagementPartners() {
      return (rootState) => rootState.data
    },
  },
```

```js
TS2349: Cannot invoke an expression whose type lacks a call signature. Type 'Selector<any, any, any>' has no compatible call signatures.
```

***

Did I miss something?

Thanks!

**Comment by d3dc** [comment]
This is just `core` typings taking precedence. My attempt to fix it is in master. I'm pretty sure we're just waiting on the next release ( fixing all the other typescript :) )

**Comment by ShMcK** [comment]
Typings seem to work in the current release, with the exception of select plugin typings. 

Inside of the select plugin, it attempts to modify the core module typings by extending the Model with a `selectors` key.

```ts
declare module '@rematch/core' {
	export interface ModelConfig<S = any, SS = S> {
		selectors?: ModelSelectorsConfig<S>,
	}
}
```

TypeScript does not seem to register type extensions if they are located in a different module.

Any suggestions on how we might proceed?

**Comment by d3dc** [comment]
@ShMcK 

> error TS2717: Subsequent property declarations ...

Doesn't this imply the compiler found the extension?

**Comment by ShMcK** [comment]
True, the compiler seems to find the type extension. but still creates conflicting typings when using `selectors`.

Unfortunately, I'm not sure what causes the problem, even after reading the TS docs under [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html).

Any ideas?

**Comment by d3dc** [comment]
I'm working on an example: https://codesandbox.io/s/l7ok4x13wq

I think it all gets found, just some things don't infer type (I'm pretty sure I attempted type information for slice, but its always `any`)...

**Comment by 19majkel94** [comment]
Looks like the combination:
```json
{
  "@rematch/core": "^1.0.5",
  "@rematch/select": "^2.0.2",
  "typescript": "^3.1.1"
}
```
has no problems with declarations merging. You just need to import the `@rematch/select` package somewhere in your project code (e.g. by creating selector plugin).

But there are a few types issues that makes it awful to use with TypeScript, e.g. the `createSelector` has wrong inferred signature when we use it like in rematch docs:
```ts
getSum() {
  return createSelector(slice, state => 
    state.items.reduce((sum, item) => sum + item.price, 0)
  );
},
```
It gets this error:
> [ts] Property 'items' does not exist on type 'Selector<CartState, void, any>'.

Although when we provide the type explicitly, it works ok:
```ts
getSum() {
  return createSelector<any, CartState, number>(slice, state => 
    state.items.reduce((sum, item) => sum + item.price, 0)
  );
}
```
I think we should create our own type for `createSelector` that will be designed for usage with `slice` and better inferring capabilities. 

And I'm missing the `ExtractRematchSelectorsFromModels` utility type from 1.0 - without this we can do `store.select.foo.bar(store.getState());` with no problem. I know it might be hard to infer the type of the param/props of the selector but at least detecting method names is needed.

**Comment by ShMcK** [comment]
Is this issue resolved as of #540? 

Released as @rematch/select@2.0.3
